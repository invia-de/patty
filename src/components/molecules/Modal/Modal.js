import React from 'react';
import PropTypes from 'prop-types';
import cx from '../../../utils/classnames';
import Overlay from '../../atoms/Overlay/Overlay';

import styles from './Modal.module.scss';

/**
 * @author [Roman Semko](mailto:roman.semko-extern@invia.de)
 * @since 0.1.0
 */
export default class Modal extends React.Component {
  constructor() {
    super();
    this.state = {
      open: false
    };
    this.onEscape = this.onEscape.bind(this);
    this.closeModal = this.closeModal.bind(this);
    this.openModal = this.openModal.bind(this);
    this.modalRef = React.createRef();
  }

  componentDidMount() {
    document.addEventListener('keydown', this.onEscape, false);
  }
  componentWillUnmount() {
    document.removeEventListener('keydown', this.onEscape, false);
  }

  render() {
    const {
      children,
      trigger,
      className,
      isStatic,
      overlayClassName
    } = this.props;
    const { open } = this.state;

    return (
      <div>
        <div className={styles.trigger} onClick={this.openModal}>
          {trigger}
        </div>
        <Overlay
          open={open}
          onClick={this.closeModal}
          isStatic={isStatic}
          className={cx(styles.overlay, overlayClassName)}
        >
          <div
            className={cx(styles.modal, isStatic && styles.static, className)}
            ref={ref => {
              this.modalRef = ref;
            }}
            onClick={e => e.stopPropagation()}
          >
            {children}
          </div>
        </Overlay>
      </div>
    );
  }

  onEscape(event) {
    if (event.keyCode === 27) {
      this.closeModal();
    }
  }

  closeModal() {
    const { onClose } = this.props;
    this.setState({ open: false });
    onClose && onClose();
  }

  openModal() {
    const { onClick, onOpen } = this.props;

    this.setState({ open: true }, () => {
      if (this.modalRef) {
        this.modalRef && this.modalRef.focus();
        this.modalRef.parentElement.style.opacity = 1;
      }
      onClick && onClick();
      onOpen && onOpen();
    });
  }
}

Modal.propTypes = {
  /** additional classNames you want to add */
  className: PropTypes.string,
  /** additional overlay classNames you want to add */
  overlayClassName: PropTypes.string,
  /** required due to accessibility */
  children: PropTypes.node.isRequired,
  /** Trigger/button. Can be null if modal is triggered pragmatically */
  trigger: PropTypes.element,
  /* triggered when the modal is opened */
  onClick: PropTypes.func,
  /* triggered when the modal is opened */
  onOpen: PropTypes.func,
  /* triggered when the modal is closed */
  onClose: PropTypes.func,
  /* prevents modal from unfolding full-screen on smaller devices */
  isStatic: PropTypes.bool
};
