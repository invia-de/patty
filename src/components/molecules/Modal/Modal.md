A simple modal that is opened, when someone clicks on the `trigger`.
The modal can be closed either by clicking out of the modal (overlay click)
or pressing the escape button.

```jsx
<Modal
  trigger={
    <button
      style={{ background: '#333', color: '#fff', border: 0, padding: '5px' }}
    >
      Open modal
    </button>
  }
>
  <div style={{ width: 480, height: 60, padding: 20, textAlign: 'center' }}>
    Hello world!
  </div>
</Modal>
```
