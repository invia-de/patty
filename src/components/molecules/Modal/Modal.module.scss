@import 'theme.scss';

$modal_background-color: $white !default;
$modal-overlay_transition: 0.3s !default;

.modal {
  background-color: $modal_background-color;

  @include respond-below(sm) {
    width: 100%;
    min-height: 100%;
  }
}

.overlay {
  opacity: 0;
  transition: $modal-overlay_transition;
}
