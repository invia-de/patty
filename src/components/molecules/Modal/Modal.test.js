import React from 'react';
import Modal from './Modal';
import { render } from 'react-testing-library';

test('Modal renders', () => {
  const modal = (
    <Modal trigger={<button>Open modal</button>}>Hello World!</Modal>
  );
  const { container } = render(modal);
  expect(container.firstChild).toMatchSnapshot();
  expect(container.firstChild.childNodes).toHaveLength(1);

  const trigger = container.firstChild.children[0];
  trigger.click();

  expect(container).toMatchSnapshot();
  expect(container.firstChild.childNodes).toHaveLength(2);

  expect(container.firstChild.childNodes[1].textContent).toEqual(
    'Hello World!'
  );
});
