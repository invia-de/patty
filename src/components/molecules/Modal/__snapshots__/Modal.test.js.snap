// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders 1`] = `
<div>
  <div
    class="trigger"
  >
    <button>
      Open modal
    </button>
  </div>
</div>
`;

exports[`Modal renders 2`] = `
<div>
  <div
    class="trigger"
  >
    <button>
      Open modal
    </button>
  </div>
  <div
    class="overlay overlay"
    style="opacity: 1;"
  >
    <div
      class="modal"
    >
      Hello World!
    </div>
  </div>
</div>
`;
